src/Codeception/Lib/Connector/Guzzle.php

```diff
@@ -107,38 +107,40 @@ protected function createResponse(Response $response)
             }
             $response->setHeader('Content-Type', $contentType);
         }
-        
+
         $headers = $response->getHeaders();
         $status = $response->getStatusCode();
-        $matches = [];
-
-        $matchesMeta = preg_match(
-            '/\<meta[^\>]+http-equiv="refresh" content="(\d*)\s*;?\s*url=(.*?)"/i',
-            $response->getBody(true),
-            $matches
-        );
+        if ($status < 300 || $status >= 400) {
+            $matches = [];
 
-        if (!$matchesMeta) {
-            // match by header
-            preg_match(
-                '~(\d*);?url=(.*)~',
-                (string)$response->getHeader('Refresh'),
+            $matchesMeta = preg_match(
+                '/\<meta[^\>]+http-equiv="refresh" content="(\d*)\s*;?\s*url=(.*?)"/i',
+                $response->getBody(true),
                 $matches
             );
-        }
 
-        if ((!empty($matches)) && (empty($matches[1]) || $matches[1] < $this->refreshMaxInterval)) {
-            $uri = $this->getAbsoluteUri($matches[2]);
-            $partsUri = parse_url($uri);
-            $partsCur = parse_url($this->getHistory()->current()->getUri());
-            foreach ($partsCur as $key => $part) {
-                if ($key === 'fragment') {
-                    continue;
-                }
-                if (!isset($partsUri[$key]) || $partsUri[$key] !== $part) {
-                    $status = 302;
-                    $headers['Location'] = $matchesMeta ? htmlspecialchars_decode($uri) : $uri;
-                    break;
+            if (!$matchesMeta) {
+                // match by header
+                preg_match(
+                    '~(\d*);?url=(.*)~',
+                    (string)$response->getHeader('Refresh'),
+                    $matches
+                );
+            }
+
+            if ((!empty($matches)) && (empty($matches[1]) || $matches[1] < $this->refreshMaxInterval)) {
+                $uri = $this->getAbsoluteUri($matches[2]);
+                $partsUri = parse_url($uri);
+                $partsCur = parse_url($this->getHistory()->current()->getUri());
+                foreach ($partsCur as $key => $part) {
+                    if ($key === 'fragment') {
+                        continue;
+                    }
+                    if (!isset($partsUri[$key]) || $partsUri[$key] !== $part) {
+                        $status = 302;
```

Comment:
What exactly have you changed?
I see many formatting changes, but it still overrides status code.